enhancements/cloud-integration/ibm/service-endpoint-dynamic-override.md

```diff
@@ -0,0 +1,127 @@
+
+## Motivation
+
+IBM Cloud requires this enhancement for their control plane replatforming efforts. OpenShift IPI for IBM Cloud will be used to deploy a cluster with critical responsibility in our managed control plane. During the genesis phase of region bringup, existing service endpoints will be used while new regional service endpoints are brought up. Once the new regional service endpoints are available, the aforementioned cluster's service endpoint configuration will need to be updated.
+
+### User Stories
+
+* As an OpenShift cluster administrator, I want to update my cluster's current service endpoint configuration to point to new service endpoints.
+ 
+### Goals
+
+* Provide an official path for IBM deployed clusters to update the infrastructure object with service endpoint override(s) that will propagate to all dependent components without further user intervention.
+
+
+
+## Proposal
+
+To realize this enhancement:
+
+* Expand API definition to support defining services + endpoints within cloud provider spec for IBM
+* Modify CCCMO so that changes are reconciled from infrastructure spec to status for IBM cloud provider and cloud config
+* Evaluate IBM components with service endpoint support and how they will react/respond to post-install service endpoint changes.
+
+### Workflow Description
+
+**cluster administrator** is a human user responsible managing an existing openshift custer deployed on IBM infrastructure.
+
+1. The cluster administrator wishes to use private IBM Cloud endpoints.
+2. The cluster administrator identifies the services that they wish to update (ie IAM and resource controller) and identifies the endpoints for these services.
+3. The cluster administrator updates the infrastructure object to contain a list of overrides where each element is the name of the service and the endpoint to use for that service. `oc edit infrastructure  -n default cluster`
+4. Once the service endpoint override update has been processed/reconciled, components can act on the change (if applicable) and use in future operations (note: may need to be restarted to pick up the change).
+
+**cccmo** is an operator responsible for watching updates to the infrastructure object and perforning updates once any value(s) are set. 
+
+1. The cccmo reconciliation loop observes that the IBMCloudPlatform spec within the infrastructure object has been set,
+2. The cccmo validates the endpoints and then updates the IBMCloudPlatformStatus and cloud config.
```

Comment:
is there anything special about the validation process?

Reply:
Validation involves ensuring the provided endpoint is a valid url using https for traffic, follows a particular pattern we expect for endpoints to follow ie `/api/v1` or `/v1` which are IBM standard, and whether the hostname exists and can be reached.

Reply:
Does this need to be done by CCMO or can it be done at admission time using something like CEL validations?

Reply:
CEL validations seem like a possible solution to validating the pattern of the provided endpoints earlier on, although we would also want to include a lookup by CCCMO to validate the endpoint exists. Is this CEL pattern desired by RH? Could you provide any other examples that have been done recently that I could look at? Thanks.

Reply:
Yes, early validation that it is in fact a valid URL is desired.

You can search the API repo for `XValidation` to find examples, but I guess from this case, we at least need to make sure that the values are valid URLs

```
// +kubebuilder:validation:XValidation:rule="isURL(self)",message="... must be a valid URL"
```


What else would we want to check? Do we need to check that the URL is secure? That a port is specified? That the Hostname follows a particular pattern?

The URL library is [documented here](https://kubernetes.io/docs/reference/using-api/cel/#kubernetes-url-library)

Reply:
The goals of validation were the host existing, the particular path following the exiting IBM convention, and that traffic for this endpoint would be using https. Besides that nothing else was required for validation.

Reply:
Makes sense, if possible, validating as much of that at admission time using CEL would provider a more immediate UX and be preferable to later errors from controllers